core/peerset/src/slots.rs

```diff
@@ -104,45 +122,88 @@ impl Slots {
 			return SlotState::MaxConnections(peer_id);
 		}
 
-		self.slots.insert(peer_id.clone(), slot_type);
+		match slot_type {
+			SlotType::Common => self.common.insert(peer_id.clone(), ()),
+			SlotType::Reserved => self.reserved.insert(peer_id.clone(), ()),
+		};
+
 		SlotState::Added(peer_id)
 	}
 
-	pub fn clear_common_slots(&mut self) -> Vec<PeerId> {
-		let slots = mem::replace(&mut self.slots, LinkedHashMap::with_capacity(self.max_slots));
-		let mut common_peers = Vec::new();
-		for (peer_id, slot_type) in slots {
-			match slot_type {
-				SlotType::Common => {
-					common_peers.push(peer_id);
-				},
-				SlotType::Reserved => {
-					self.slots.insert(peer_id, slot_type);
-				},
-			}
+	/// Pops the oldest peer from the list.
+	pub fn pop_peer(&mut self, reserved_only: bool) -> Option<(PeerId, SlotType)> {
+		if let Some((peer_id, _)) = self.reserved.pop_front() {
+			return Some((peer_id, SlotType::Reserved));
```

Comment:
It seems kind of strange to pop the reserved peers first. wouldn't we rather cycle out the common ones earlier?

Reply:
I see where ur confusion comes from. `pop_peer` is only used for `discovered` list. We pop a peer that is later inserted into `out_slots` or `in_slots`